### DISCUSSION TITLE
Are reaction callbacks automatically run in (trans)actions?

### DISCUSSION THREAD

**Discussion by Dorosty** [mobxjs/mobx#3376] [Q&A] [answered]
Apologies in advance if this has been answered/documented before. I couldn't find an answer in the docs.

Are reaction callbacks automatically run in (trans)actions? What I mean is, is this necessary?
```js
autorun(action(() => { /****/ }))
reaction(getDataFromState, action(data => { /****/ }))
when(predicate, action(() => { /****/ }))
```

Or does this suffice?
```js
autorun(() => { /****/ })
reaction(getDataFromState, data => { /****/ })
when(predicate, () => { /****/ })
```

**Comment by urugator** [comment, marked as answer]
> Are reaction callbacks automatically run in (trans)actions?

yes

> What I mean is, is this necessary?

no

> Or does this suffice?

yes

However you should avoid mutating state in reactions, therefore it shouldn't matter most of the time.

**Reply by Dorosty** [reply]
Thanks! Is the same thing true for generator functions? If I pass one as a callback will it automatically get wrapped with `flow`?

**Reply by urugator** [reply]
No, interesting idea though.

**Reply by Dorosty** [reply]
Last question. Are callbacks passed to `onBecomeObserved` and `onBecomeUnobserved` also run as actions?

**Reply by urugator** [reply]
No, `observe`/`intercept` callbacks neither.

**Reply by urugator** [reply]
Btw you can inspect `_getGlobalState().inBatch`, non-zero value means you're in batch (usually in action, but not necessary).
